### PR TITLE
ENH: Add parameter file support for TransformOutputFileNameExtensions

### PR DESCRIPTION
**Pull request by N-Dekker**
Allows specifying output file types for transform files by their file name extension, for example:

    (TransformOutputFileNameExtensions ".h5" ".mat" ".tfm")

### CODE REVIEW

**Comment by mstaring** [comment]
this looks good :-)
Does it work with elastix transformation objects as well?

**Comment by N-Dekker** [comment]
> this looks good :-)

@mstaring Thanks! So you like the feature of allowing the user to specify multiple output file types, by their file name extensions (".h5", ".mat", ".tfm")?

> Does it work with elastix transformation objects as well?

Needs to be tested!  As far as I can see, `TransformBase::WriteToFile(const ParametersType &)` is overridden by 15 elastix transform classes:

    AdvancedAffineTransform
    AdvancedBSplineTransform
    AffineDTITransform
    AffineLogStackTransform
    AffineLogTransform
    BSplineStackTransform
    DeformationFieldTransform
    EulerStackTransform
    EulerTransform
    MultiBSplineTransformWithNormal
    RecursiveBSplineTransform
    SimilarityTransform
    SplineKernelTransform
    TranslationStackTransform
    WeightedCombinationTransform

Does `itk::TransformFileWriter` do the job for all of them correctly? I do not know.

Another thing I wonder: what about the `ParametersType` argument (`param`) of `WriteToFile(const ParametersType &)`? It seems redundant, because it seems that all the transform data is already in the transform object. But I'm not sure.

There is a `WriteToFile(void)` overload that just calls the other one, as in  `WriteToFile(GetAsITKBaseType()->GetParameters())`. But it isn't being used anywhere! Why not? Instead, `ElastixTemplate::CreateTransformParameterFile` calls `GetElxTransformBase()->WriteToFile(this->GetElxOptimizerBase()->GetAsITKBaseType()->GetCurrentPosition())`.

**Comment by N-Dekker** [comment]
TODO Try-catch

**Comment by N-Dekker** [comment]
Hmmm.... it appears that even a very simple TFM file that is generated by elastix this way (this pull request) cannot be read into 3D Slicer: 

    #Insight Transform File V1.0
    #Transform 0
    Transform: TranslationTransformElastix_double_3_3
    Parameters: 0.9908445154342919 1.9923078126157387 2.992131055892518
    FixedParameters: 

3D Slicer error log says:
> Could not create an instance of "TranslationTransformElastix_double_3_3"
The usual cause of this error is not registering the transform with TransformFactory
Currently registered Transforms:

**Comment by mstaring** [comment]
what if you manually edit the name of the transform to TranslationTransform ? (or the name that itk uses in their code)

**Comment by N-Dekker** [comment]
> what if you manually edit the name of the transform to TranslationTransform ? (or the name that itk uses in their code)

It appears that data loading is very fragile. I just manually edited, so that it now says "Transform: AffineTransform_double_3_3", because that is what 3D Slider puts into its generated TFM files. The tried to load the file, and 3D Slider just crashed! (Exit program!)

Now I added "FixedParameters: 0 0 0" and it produces error messages: 

> Error setting parameters: parameters array size (3) is less than expected  (NInputDimensions * NOutputDimensions + NOutputDimensions)  (3 * 3 + 3 = 12)
> Failed to read transform from file: D:/X/data/3D/out/TransformParameters.0_edited.tfm

To be continued...

**Comment by mstaring** [comment]
yes, for affine you have the 3x3 rotation matrix and 3 offsets (translations). what if for fixedparams you use:

1 0 0  0 1 0 0 0 1 0 0 0
?

**Comment by mstaring** [comment]
the crash perhaps you can report to the slicer people, perhaps we can collaborate on this. They have been interested in this functionality in the past

**Comment by N-Dekker** [comment]
For the record, ITK IO supports the following 36 transform types out of the box:

    AffineTransform
    AzimuthElevationToCartesianTransform
    BSplineDeformableTransform
    BSplineSmoothingOnUpdateDisplacementFieldTransform
    BSplineTransform
    CenteredAffineTransform
    CenteredEuler3DTransform
    CenteredRigid2DTransform
    CenteredSimilarity2DTransform
    CompositeTransform
    ConstantVelocityFieldTransform
    DisplacementFieldTransform
    Euler2DTransform
    Euler3DTransform
    FixedCenterOfRotationAffineTransform
    GaussianExponentialDiffeomorphicTransform
    GaussianSmoothingOnUpdateDisplacementFieldTransform
    GaussianSmoothingOnUpdateTimeVaryingVelocityFieldTransform
    IdentityTransform
    QuaternionRigidTransform
    Rigid2DTransform
    Rigid3DPerspectiveTransform
    Rigid3DTransform
    ScalableAffineTransform
    ScaleLogarithmicTransform
    ScaleSkewVersor3DTransform
    ScaleTransform
    ScaleVersor3DTransform
    Similarity2DTransform
    Similarity3DTransform
    TimeVaryingBSplineVelocityFieldTransform
    TimeVaryingVelocityFieldTransform
    TranslationTransform
    VelocityFieldTransform
    VersorRigid3DTransform
    VersorTransform

For various input and output dimensionalities (2-D, 3-D, 4-D, but only a limited number of combinations), any combination both for `float` and `double`.  Total number of combinations: 116. As listed by `itk::TransformFactoryBase::GetFactory()->GetClassOverrideWithNames()`.

**Comment by N-Dekker** [comment]
For the record, 17 transform classes in elastix\Components\Transforms have an `elxClassNameMacro` call:

directory | elxGetClassName
-- | --
AdvancedAffineTransform | AffineTransform
AdvancedBSplineTransform | BSplineTransform
AffineDTITransform | AffineDTITransform
AffineLogStackTransform | AffineLogStackTransform
AffineLogTransform | AffineLogTransform
BSplineDeformableTransformWithDiffusion | BSplineTransformWithDiffusion
BSplineStackTransform | BSplineStackTransform
DeformationFieldTransform | DeformationFieldTransform
EulerStackTransform | EulerStackTransform
EulerTransform | EulerTransform
MultiBSplineTransformWithNormal | MultiBSplineTransformWithNormal
RecursiveBSplineTransform | RecursiveBSplineTransform
SimilarityTransform | SimilarityTransform
SplineKernelTransform | SplineKernelTransform
TranslationStackTransform | TranslationStackTransform
TranslationTransform | TranslationTransform
WeightedCombinationTransform | WeightedCombinationTransform

**Comment by N-Dekker** [comment]
For the record, ITK Transform classes and their number of parameters (both non-fixed "Parameters" and "FixedParameters"), by default:

n | ITK class | # Parameters | # FixedParameters
-- | -- | -- | --
1 | AffineTransform<double,2> | 6 | 2
2 | AffineTransform<double,3> | 12 | 3
3 | AffineTransform<double,4> | 20 | 4
4 | AzimuthElevationToCartesianTransform<double,3> | 12 | 3
5 | BSplineDeformableTransform<double,2,2> | 0 | 10
6 | BSplineDeformableTransform<double,3,3> | 0 | 18
7 | BSplineSmoothingOnUpdateDisplacementFieldTransform<double,2> | 0 | 10
8 | BSplineSmoothingOnUpdateDisplacementFieldTransform<double,3> | 0 | 18
9 | BSplineTransform<double,2,3> | 32 | 10
10 | BSplineTransform<double,3,3> | 192 | 18
11 | CenteredAffineTransform<double,2> | 8 | 2
12 | CenteredAffineTransform<double,3> | 15 | 3
13 | CenteredEuler3DTransform<double> | 9 | 4
14 | CenteredRigid2DTransform<double> | 5 | 2
15 | CenteredSimilarity2DTransform<double> | 6 | 2
16 | CompositeTransform<double,2> | 0 | 0
17 | CompositeTransform<double,3> | 0 | 0
18 | CompositeTransform<double,4> | 0 | 0
19 | ConstantVelocityFieldTransform<double,2> | 0 | 10
20 | ConstantVelocityFieldTransform<double,3> | 0 | 18
21 | DisplacementFieldTransform<double,2> | 0 | 10
22 | DisplacementFieldTransform<double,3> | 0 | 18
23 | Euler2DTransform<double> | 3 | 2
24 | Euler3DTransform<double> | 6 | 4
25 | FixedCenterOfRotationAffineTransform<double,3> | 15 | 3
26 | GaussianExponentialDiffeomorphicTransform<double,2> | 0 | 10
27 | GaussianExponentialDiffeomorphicTransform<double,3> | 0 | 18
28 | GaussianSmoothingOnUpdateDisplacementFieldTransform<double,2> | 0 | 10
29 | GaussianSmoothingOnUpdateDisplacementFieldTransform<double,3> | 0 | 18
30 | GaussianSmoothingOnUpdateTimeVaryingVelocityFieldTransform<double,2> | 0 | 18
31 | GaussianSmoothingOnUpdateTimeVaryingVelocityFieldTransform<double,3> | 0 | 28
32 | IdentityTransform<double,2> | 0 | 0
33 | IdentityTransform<double,3> | 0 | 0
34 | IdentityTransform<double,4> | 0 | 0
35 | QuaternionRigidTransform<double> | 7 | 3
36 | Rigid2DTransform<double> | 3 | 2
37 | Rigid3DPerspectiveTransform<double> | 6 | 0
38 | v3::Rigid3DTransform<double> | 12 | 3
39 | ScalableAffineTransform<double,3> | 12 | 3
40 | ScaleLogarithmicTransform<double,3> | 3 | 3
41 | ScaleSkewVersor3DTransform<double> | 15 | 3
42 | ScaleTransform<double,2> | 2 | 2
43 | ScaleTransform<double,3> | 3 | 3
44 | ScaleTransform<double,4> | 4 | 4
45 | ScaleVersor3DTransform<double> | 9 | 3
46 | Similarity2DTransform<double> | 4 | 2
47 | Similarity3DTransform<double> | 7 | 3
48 | TimeVaryingBSplineVelocityFieldTransform<double,2> | 0 | 18
49 | TimeVaryingBSplineVelocityFieldTransform<double,3> | 0 | 28
50 | TimeVaryingVelocityFieldTransform<double,2> | 0 | 18
51 | TimeVaryingVelocityFieldTransform<double,3> | 0 | 28
52 | TranslationTransform<double,2> | 2 | 0
53 | TranslationTransform<double,3> | 3 | 0
54 | TranslationTransform<double,4> | 4 | 0
55 | VelocityFieldTransform<double,2> | 0 | 18
56 | VelocityFieldTransform<double,3> | 0 | 28
57 | VersorRigid3DTransform<double> | 6 | 3
58 | VersorTransform<double> | 3 | 3

File: Core/ComponentBaseClasses/elxTransformBase.hxx [hunk 1]
```diff
@@ -653,6 +656,78 @@ TransformBase<TElastix>::WriteToFile(const ParametersType & param) const
...
+    itk::TransformFactoryBase::GetFactory();
+
+    const auto transformObject = [this]() -> itk::Object::ConstPointer {
+      const auto thisAsITKBase = this->GetAsITKBaseType();
+      assert(thisAsITKBase != nullptr);
+
+      // The string returned by elxGetClassName() corresponds to ITK's GetNameOfClass(),
+      // for AffineTransform, BSplineTransform, and TranslationTransform.
+      // Note that for EulerTransform and SimilarityTransform, ITK has "2D"
+      // or "3D" inserted in the class name.
```
Comment (N-Dekker) [inline Core/ComponentBaseClasses/elxTransformBase.hxx:677]: Assumes a correspondence between parameters of elastix transforms and ITK transforms, based on elastic class-name and the ITK name-of-class, for AffineTransform, BSplineTransform, TranslationTransform, EulerTransform (for 2D and 3D), SimilarityTransform (for 2D and 3D).

**Comment by N-Dekker** [comment]
> the crash perhaps you can report to the slicer people, perhaps we can collaborate on this. They have been interested in this functionality in the past

The crash (which happens when the FixedParameters in the input file is incomplete) appears in `itk::MatrixOffsetTransformBase::SetFixedParameters(const FixedParametersType &)`, at https://github.com/InsightSoftwareConsortium/ITK/blob/v5.1.1/Modules/Core/Transform/include/itkMatrixOffsetTransformBase.hxx#L491 It affects all ITK classes that derive from `MatrixOffsetTransformBase`, specifically:

    AffineTransform
    AzimuthElevationToCartesianTransform
    CenteredAffineTransform
    Euler2DTransform
    FixedCenterOfRotationAffineTransform
    QuaternionRigidTransform
    Rigid2DTransform
    Rigid3DTransform
    ScalableAffineTransform
    ScaleLogarithmicTransform
    ScaleSkewVersor3DTransform
    ScaleTransform
    ScaleVersor3DTransform
    Similarity2DTransform
    Similarity3DTransform
    VersorRigid3DTransform
    VersorTransform

**Comment by N-Dekker** [comment]
For the record, the following ITK transforms do not have a lossless write-and-read-back round-trip, for a default-constructed transform:

    ConstantVelocityFieldTransform
    GaussianExponentialDiffeomorphicTransform
    GaussianSmoothingOnUpdateTimeVaryingVelocityFieldTransform
    TimeVaryingBSplineVelocityFieldTransform
    TimeVaryingVelocityFieldTransform
    VelocityFieldTransform 

While trying to read the transform back in, they encounter the folowing error, from `ImageBase<>::ComputeIndexToPhysicalPointMatrices()`: 

> ITK\Modules\Core\Common\include\itkImageBase.hxx:177:
> itk::ERROR: itk::ERROR: Image(00000216A7674990): A spacing of 0 is not allowed: Spacing is [0, 0...

Call stack:

    ImageBase::ComputeIndexToPhysicalPointMatrices() Line 177
    Image<itk::Vector>::ComputeIndexToPhysicalPointMatrices() Line 142
    ImageBase::SetSpacing(const itk::Vector & ) Line 104
    ConstantVelocityFieldTransform::SetFixedParameters(const itk::OptimizerParameters & ) Line 199
    TxtTransformIOTemplate::Read() Line 221
    TransformFileReaderTemplate::Update() Line 139

**Comment by N-Dekker** [comment]
Will merge this pull request, as it offers the basics for writing transform files. However, each individual transform type still needs to be tested more thoroughly, which will possibly suggest some code changes in individual transform types.

**Pull request merged**